Adding tags to Sets

Spijkmans opened the pull request:
Changes:
- Added tags for Active and Inactive states, listing the labels of the states respectively (per set);
- Added tags for number of Active  and Inactive states per set;
- Added tag for total number of states per set;
- Added option in settings to change/set the separator string;
- Updated the locale json's as needed

Issues:
- The 'ng-attr-placeholder=="{{ set.separator }}"' is not updated after changing the separator

@eriks5 Please review this pull request. I'm kinda curious if i did it correctly, as this is my first pull request.

Spijkmans committed:
Adding tags to Sets

Changes:
- Added tag for Active and Inactive states, listing the labels of the states respectively (per set);
- Added tags for number of Active  and Inactive states per set;
- Added tag for total number of states per set;
- Added option in settings to change/set the separator string;
- Updated the locale json's as needed

Spijkmans committed:
Fixed toggle first state

Fixed a bug where (re)starting the Sets app would toggle the first state.

Spijkmans committed:
Fixed bug in first state

Fixed a bug where the first state would toggle on (re)starting the Sets app

Spijkmans committed:
Consolidating tag updates

In this update:
- Updating the tags without triggering state changes
- Consolidating tag creation and update

Spijkmans committed:
Merge branch 'master' of https://github.com/Spijkmans/org.scheffers.homey.sets

Spijkmans committed:
Consolidating tags update and creation

- Consolidated the update/creation of the tags in one function
- Changed tag creation so it doesn't trigger state changes

Spijkmans committed:
Update .gitignore

Spijkmans committed:
Merge branch 'master' of https://github.com/Spijkmans/org.scheffers.homey.sets

Spijkmans committed:
Settings page updated after separator change

Fixed: The settings page is now updated, async, after changing the separator

Spijkmans commented:
@eriks5 I think the pull-request is ready for you to test.

Spijkmans commented:
@eriks5 did you have a look at my proposed code changes?
Just curious about your reaction...

eriks5 commented:
No sorry haven't had time yet to dive into it properly. It's been years since I did Homey app development, my development environment doesn't work anymore so I need to fix that before I can properly look at this.

I did do a quick scan. One thing I did notice is that you use global variables (app.js lines 30-40), which is generally a bad idea, and the Homey SDK 3 changelog (https://apps-sdk-v3.developer.athom.com/tutorial-SDK%20Changelog.html) explicitly says it may cause issues in the future:

> For the same reason we have moved away from "polluting the global namespace" we urge you to define all variables as properties on your App, Driver or Device instances. Your global scope (anything that isn't inside of a class) should only contain constants (their values shouldn't be changed). Relying on global state may cause issues for your app in the future.

So that's something that definitely needs to be fixed.

Also, the README.md needs to be updated to document this feature before it can be released.

Spijkmans committed:
Fixes and updates

Updated the README.md to reflect the changes in this release.

Fixed:
 - prefix vars and label vars changed to const.
 - moved setsAllTokens and setsAllTokensIndex from Global namespace to SetsApp class.

Spijkmans commented:
@eriks5 you correctly noted the issues with Global vars and the need to update the README.md.
I just uploaded an updated app.js and updated README.md.

Please take a look at these changes and let me know what you think.

Happy Easter

Spijkmans commented:
Some afterthoughts;
- Should I increase the version number each time I commit? Or maybe I shuold use build versions?
- The Homey CLI gives warnings when Validating the app;
`Warning: flow.triggers[not_none_active].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.triggers[none_active].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.triggers[none_active_changed].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.triggers[all_active].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.triggers[not_all_active].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.triggers[all_active_changed].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.triggers[state_set].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.triggers[state_reset].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.triggers[change].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.conditions[none_active].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.conditions[all_active].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.conditions[is_active].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.actions[activate_state].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.actions[activate_temp_state].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.actions[activate_delayed].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.actions[deactivate_state].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.actions[activate_all].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.actions[deactivate_all].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.
Warning: flow.actions[activate_one].titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.`
I'm not sure how to fix this. Simply updating the app.json doesn't seem to help...

eriks5 commented:
You generally shouldn't touch the version number in a pull request actually. Version number should bump once on the merge.

I'll look into the warnings when I look at the rest.

eriks5 commented:
I've looked into it. Looks good. 

However I have some doubts about the changes to the settings page. What's the use-case of having different separators for different sets? I'm afraid having a configurable separator for each set clutters the settings page too much right now.

In fact, why would the separator have to be configurable at all? Wouldn't the default be good enough for (almost) everyone?

Spijkmans commented:
Glad you found it 'looking good 😄'.

About the configurable separator, I wasn't sure if a single separator would work for different languages and/or different situations.
I mean maybe in one language one would expect a '/' or a '-' as separator while an other language a ', ' (like in dutch) is more or less expected.

However I do agree it does clutter the settings page. And I struggled a bit with that.

So maybe we should settle with a fixed separator for now.
